./main.go:56:3: undefined: gosyncmodules.CheckPerm

build failures
```
➜  ADtoLDAP git:(master) $ GOOS=linux GOARCH=amd64 go build .
./main.go:56:3: undefined: gosyncmodules.CheckPerm
```
go version:
```go version go1.11.4 darwin/amd64```

ADtoLDAP version is master

This works with native build, but this doesn't work with cross-compiling from mac to linux because of CGO and I see failures even with `CGO_ENABLED=1 `.

The error I see when cross compiling on mac is
```
ld: warning: ignoring file /var/folders/y4/qf97rh1x66148z9k7f6m6vc40000gp/T/go-link-754692567/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/y4/qf97rh1x66148z9k7f6m6vc40000gp/T/go-link-754692567/go.o
Undefined symbols for architecture x86_64:
  "__cgo_topofstack", referenced from:
      __cgo_3298b262a8f6_Cfunc_mygetgrgid_r in 000001.o
      __cgo_3298b262a8f6_Cfunc_mygetgrnam_r in 000001.o
      __cgo_3298b262a8f6_Cfunc_mygetpwnam_r in 000001.o
      __cgo_3298b262a8f6_Cfunc_mygetpwuid_r in 000001.o
      __cgo_3298b262a8f6_Cfunc_realloc in 000001.o
      __cgo_3298b262a8f6_Cfunc_sysconf in 000001.o
      __cgo_3298b262a8f6_Cfunc_mygetgrouplist in 000002.o
      ...
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


```


 Will look into it. This doesn't need cgo now.

--- a/gosyncmodules/CheckPerm.go
+++ b/gosyncmodules/CheckPerm.go
@@ -5,51 +5,56 @@ import (
 	"os"
 	"os/user"
 	"strconv"
-	"unsafe"
+	"syscall"
 )
 
-//#include <sys/stat.h>
-//#include <stdlib.h>
-import "C"
+// Constants S_IRGRP, S_IWGRP, S_IXGRP, S_IROTH, S_IWOTH and S_IXOTH values are taken straight from /stat.h.
+// Ref: https://github.com/torvalds/linux/blob/master/include/uapi/linux/stat.h
+const (
+	IRGRP = 0000040
+	IWGRP = 0000020
+	IXGRP = 0000010
+	IROTH = 0000004
+	IWOTH = 0000002
+	IXOTH = 0000001
+)
 
+// CheckPerm checks permission of the passed filename(string) and panics if group or others have > read permissions.
 func CheckPerm(filename string) {
-	current_uid := os.Getuid()
-	current_gid := os.Getgid()
-	username, err := user.LookupId(strconv.Itoa(current_uid))
-	var current_username string
-	var current_groupname string
+	currentUID := os.Getuid()
+	currentGID := os.Getgid()
+	username, err := user.LookupId(strconv.Itoa(currentUID))
+	var currentUserName string
+	var currentGroupName string
 	if err != nil {
-		current_username = "<Couldn't lookup username>"
+		currentUserName = "<Couldn't lookup username>"
 	}
-	groupname, err := user.LookupGroupId(strconv.Itoa(current_gid))
+	groupname, err := user.LookupGroupId(strconv.Itoa(currentGID))
 	if err != nil {
-		current_groupname = "<Couldn't lookup groupname>"
+		currentGroupName = "<Couldn't lookup groupname>"
 
 	}
-	current_username = username.Username
-	current_groupname = groupname.Name
+	currentUserName = username.Username
+	currentGroupName = groupname.Name
 
-	logger.Infoln("using cgo to perform security check on ", filename)
-	statstruct := C.stat //stat struct from C
-	logger.Infoln("Initiated stat struct")
-	path := C.CString(filename)
-	logger.Infoln("Converted native string to C.CString")
-	st := *(*C.struct_stat)(unsafe.Pointer(statstruct)) //Casting unsafe pointer to C.struct_stat
-	logger.Infoln("Casting unsafe.Pointer(stat) to *(*C.struct_stat)")
-	defer C.free(unsafe.Pointer(path)) //free the C.CString that is created in heap.
-	C.stat(path, &st)
-	uid := st.st_uid
-	gid := st.st_gid
-	if int(uid) != current_uid || int(gid) != current_gid {
-		fmt.Printf("%s not owned by uid=%d(%s) or gid=%d(%s). Do chown %d:%d %s\n", filename, current_uid, current_username, current_gid, current_groupname, current_uid, current_gid, filename)
-		logger.Infof("%s not owned by uid=%d(%s) or gid=%d(%s). Do chown %d:%d %s\n", filename, current_uid, current_username, current_gid, current_groupname, current_uid, current_gid, filename)
+	fileStat, err := os.Stat(filename)
+	if err != nil {
+		panic(err)
+	}
+	filestatSys := fileStat.Sys().(*syscall.Stat_t)
+	uid := filestatSys.Uid
+	gid := filestatSys.Gid
+	if int(uid) != currentUID || int(gid) != currentGID {
+		fmt.Printf("%s not owned by uid=%d(%s) or gid=%d(%s). Do chown %d:%d %s\n", filename, currentUID, currentUserName, currentGID, currentGroupName, currentUID, currentGID, filename)
+		logger.Errorf("%s not owned by uid=%d(%s) or gid=%d(%s). Do chown %d:%d %s\n", filename, currentUID, currentUserName, currentGID, currentGroupName, currentUID, currentGID, filename)
 		os.Exit(1)
 	}
-	if st.st_mode&C.S_IRGRP > 0 || st.st_mode&C.S_IWGRP > 0 || st.st_mode&C.S_IXGRP > 0 ||
-		st.st_mode&C.S_IROTH > 0 || st.st_mode&C.S_IWOTH > 0 || st.st_mode&C.S_IXOTH > 0 {
+	if filestatSys.Mode&IRGRP > 0 || filestatSys.Mode&IWGRP > 0 || filestatSys.Mode&IXGRP > 0 ||
+	   filestatSys.Mode&IROTH > 0 || filestatSys.Mode&IWOTH > 0 || filestatSys.Mode&IXOTH > 0 {
 		fmt.Println(filename, "file permission too broad, make it non-readable to groups and others.")
-		logger.Infoln(filename, "file permission too broad, make it non-readable to groups and others.")
+		logger.Errorln(filename, "file permission too broad, make it non-readable to groups and others.")
 		os.Exit(1)
 	}
 	logger.Infoln("File permission looks secure")
 }
+

